### DISCUSSION TITLE
Which field uniquely identifies a state? The "id" or the "name"?

### DISCUSSION THREAD

**Discussion by giorgosera** [serverlessworkflow/specification#664] [Q&A] [answered]
Hi there, which of the two fields `id` or `name` uniquely identifies a state? 

In the JSON schema [here](https://serverlessworkflow.io/schemas/latest/workflow.json) I can see that there is an `id` field on a state and the description says `Unique State id`. and for the `name` it just says `State name`.

However, when I go to the spec docs [here](https://github.com/serverlessworkflow/specification/blob/main/specification.md#Event-State) I can see that the `name` description says "Unique state name". Also, there is no mention of the `id` field there.

- So which one should be used to uniquely identify a state?
- Also, is it better to look at the JSON schema for an up-to-date documentation of the fields?

Thank you very much

**Comment by cdavernas** [comment, marked as answer]
@giorgosera My understanding of it is that state `id` is an obsolete artifact of a previous version of the spec. In Synapse and other runtimes, we uniquely identify states thanks to their `name`, which is a required property, and avoid the useless overhead for users that is to set the `id` property too, which is therefore useless.

@tsurdilo Can you please confirm/infirm? Shouldn't we do a PR for getting rid of that confusing property, anyways?

**Reply by giorgosera** [reply]
Thanks @cdavernas. It's clear. 

Does the same hold for the workflow `id` and `name`? Or in that case `id` is the unique identifier?

**Reply by cdavernas** [reply]
Not in my understanding, no. On Synapse, we use the `id` as a unique identifier for a workflow, whereas the name is used solely for display purpose/user readability. But I guess you could use the name to achieve the same.

But this does raise a question: why not apply the id/name pair to each concept? @tsurdilo @ricardozanini @JBBianchi @antmendoza  What's your take on that?

**Reply by JBBianchi** [reply]
Maybe some answers can be found in the related issues, like: https://github.com/serverlessworkflow/specification/issues/463 https://github.com/serverlessworkflow/specification/issues/464 and https://github.com/serverlessworkflow/specification/pull/585

**Reply by tsurdilo** [reply]
I think for the most part we can remove state id field as we require state name to be unique. 
Do wonder tho if this is actually useful for error handling (the updates we are doing) because we do not require unique state names between multiple workflows (parent->subflows), and in those cases having the id might be useful, but not sure yet will think about it.

**Reply by tsurdilo** [reply]
For workflows the id/key should be the unique identifier. Name is not a required field and cannot be used for uniqueness imo

**Reply by tsurdilo** [reply]
| why not apply the id/name pair to each concept?

I'm not sure, could you give more info on this? I think what's currently maybe lacking is ability for users to define domain-specific id.
Typically you have a workflow id (business id which we tried to do w/ the key param) and a workflow type which would denote the actual workflow type thats being invoked/executed. In this approach the workflow type would be unique across a single deployment. The runtime then would make the rule if multiple business level ids would be allowed to execute at the same time (governance rules for business level ids). wdyt?

**Reply by cdavernas** [reply]
> I think for the most part we can remove state id field as we require state name to be unique.
... because we do not require unique state names

State names should be unique, so no problem here IMHO ;)

**Reply by cdavernas** [reply]
>  I think what's currently maybe lacking is ability for users to define domain-specific id.

This is what `key` is for. As was demonstrated by @mswiderski in another issue.

**Reply by cdavernas** [reply]
> I'm not sure, could you give more info on this?

Drop it. It's stupid. it was the idea of adding an id for each component: states, actions, ... But that has no added value and complexifies the definition.

I would however enforce, as we do with Synapse, the requirement of naming all nameable components (conditions, actions, etc.) => this ensures a proper readability / displayability of the definition. WDYT @tsurdilo ?

**Reply by tsurdilo** [reply]
Yeah key is ok, only thing that maybe we don't specify is how user sets the key, not in the workflow definition itself but from either configuration (client apis in runtime) or workflow data input. So i don't really think we solved the "buesiness level id" ..but could be wrong

**Reply by cdavernas** [reply]
> Yeah key is ok, only thing that maybe we don't specify is how user sets the key, not in the workflow definition itself but from either configuration (client apisin runtime) or workflow data input.

To my understanding, the `key` should only be defined by workflow data input thanks to an expression. It should be a requirement, if runtimes want to use it, in order to ensure portability of written flows accross implementations.

**Reply by tsurdilo** [reply]
got it, we then should add example of that in our spec doc

**Reply by tsurdilo** [reply]
also think that we should add either as top-level prop or extension (supplied by us or runtimes i dont care) ability for user to say what should happen if multiple instsances with same business level id (key) are tried to be created by client

(what should happen in terms of "allow" "not allow", "???")

**Reply by cdavernas** [reply]
> also think that we should add either as top-level prop or extension (supplied by us or runtimes i dont care) ability for user to say what should happen if multiple instsances with same business level id (key) are tried to be created by client

@tsurdilo IMHO, key should be unique if specified, thus other instances with similar key should fault. IMHO, this should be the only allowed behavior, otherwise there's no purpose to using the key

**Reply by ricardozanini** [reply]
> This is what key is for. As was demonstrated by @mswiderski in another issue.

@cdavernas do you have the link for this issue? We are currently ignoring `key` in our implementation. I mean, if `id` is not defined, it's the identification of the workflow definition. But rather than that, it means nothing to us internally.

**Reply by tsurdilo** [reply]
@cdavernas lets talk offline or in weekly meet. think of use case where you have "entity" workflows meaning the execution can represent an atomic operation and user does not want to have more than one wf instance ood this entity. another use case is completion, where user only wants to allow an entiry to be processed if previous attempt to do so had failed
